### PR TITLE
Improve comparison and equality

### PR DESCRIPTION
**Pull request by knocte**
/cc @joemphilips

### CODE REVIEW

**Comment by nopara73** [comment]
@lontivero would you mind to review it? Smells a bit dangerous.

File: NBitcoin/UInt2561.cs [hunk 1]
```diff
@@ -1,11 +1,13 @@
 ﻿
```
Comment (joemphilips) [inline NBitcoin/UInt2561.cs]: For uint256 and uint160, this branch and master have both implemented IEquatable and IComparable,
both seems to work fine. But lets just have minimum change here. So would you remove this file from this PR? @knocte 

It should be no functional difference.

Comment (knocte) [reply NBitcoin/UInt2561.cs]: done

File: NBitcoin/PubKey.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@ public enum ScriptPubKeyType
 		/// </summary>
 		SegwitP2SH
 	}
-	public class PubKey : IBitcoinSerializable, IDestination, IComparable<PubKey>, IEquatable<PubKey>
+	public class PubKey : IBitcoinSerializable, IDestination, IComparable<PubKey>, IEquatable<PubKey>, IComparable
```
Comment (nopara73) [inline NBitcoin/PubKey.cs]: `IComparable` makes no sense for `PubKey`.

File: NBitcoin/Transaction.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 namespace NBitcoin
 {
-	public class OutPoint : IBitcoinSerializable
+	public class OutPoint : IBitcoinSerializable, IComparable, IComparable<OutPoint>, IEquatable<OutPoint>
```
Comment (nopara73) [inline NBitcoin/Transaction.cs]: `IComparable` makes no sense for `OutPoint` outpoint, unless you want to implement @kristovatla's BIP69: https://github.com/bitcoin/bips/blob/master/bip-0069.mediawiki

File: NBitcoin/Transaction.cs [hunk 4]
```diff
@@ -200,10 +193,32 @@ public override string ToString()
...
+			if (other is null)
+				return false;
+			return other == this;
+		}
+
 	}
 
 
-	public class TxIn : IBitcoinSerializable
+	public class TxIn : IBitcoinSerializable, IEquatable<TxIn>
```
Comment (nopara73) [inline NBitcoin/Transaction.cs]: `IComparable` makes no sense for `TxIn` outpoint, unless you want to implement @kristovatla's BIP69: https://github.com/bitcoin/bips/blob/master/bip-0069.mediawiki

File: NBitcoin/Transaction.cs (hunk 3 above)
Comment (joemphilips) [reply NBitcoin/Transaction.cs]: I implemented these interface just for use of convenience on my code. (To use in F# collection types) And the actual logic does not really matter. So yes, I think modifying these methods to follow BIP69 is the right way to go.

File: NBitcoin/PubKey.cs (hunk 2 above)
Comment (joemphilips) [reply NBitcoin/PubKey.cs]: I think it makes sense to follow [bip67](https://github.com/bitcoin/bips/blob/master/bip-0067.mediawiki)

**Comment by joemphilips** [comment]
@knocte Would you mind adding test vectors from bip67 and bip69 ?
Unless someone have an objection to make these objects comparable in the first place.

**Comment by knocte** [comment]
> @knocte Would you mind adding test vectors from bip67 and bip69 ?

Well, if comparison is something that helps implementation of some BIPs, I would say this change has value in and of itself. Now, if someone were to implement those BIPs in particular and finds issues, he can always write tests that cover those issues, right? What I mean is, can we just mark the Compare() methods with a //TODO or a //FIXME comment saying "add tests"? It's not that I'm lazy, it's that I don't want to get deep in yak-shaving, you know. My main motivation is that DotNetLightning stops using a fork of NBitcoin, so that consumers of DotNetLightning which were already consumers of upstream NBitcoin don't find assembly-loading conflicts.

File: NBitcoin/PubKey.cs (hunk 2 above)
Comment (nopara73) [reply NBitcoin/PubKey.cs]: ACK.

**Comment by lontivero** [comment]
Concept NACK. 
It assumes that `OutPoint` and `TxIn` are comparable and that there is only one official/correct way to do it, I mean, it assumes they should be always compared in the way that bip67 and bip69 specifies, what makes sense only in one very specific use case (the one @joemphilips needs for his project).

File: NBitcoin/Transaction.cs (hunk 4 above)
Comment (NicolasDorier) [reply NBitcoin/Transaction.cs]: NACK on this one, `TxIn` can be subclassed by shitcoins so you can't assume you know equality.

**Comment by joemphilips** [comment]
@knocte  You can close this now. I made a wrapper class on my side.

FYI, I don't really agree with @lontivero  with this one. I think having default ordering for PubKey TxOut, and OutPoint is important for anonymity (I agree with Nicolas about TxIn though) . Especially when not every devs are aware of the existence of those BIPs. And developers can define custom comparator anytime when they need other way of sorting.

But anyway, it is not a big deal for me. someone can create another PR if he thinks this is necessary.

**Pull request closed without merging**